crates/amaru/src/stages/pull_p2p.rs

```diff
@@ -0,0 +1,348 @@
+// Copyright 2024 PRAGMA
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::send;
+use amaru_consensus::consensus::ChainSyncEvent;
+use amaru_kernel::Point;
+use gasket::framework::*;
+use pallas_network2::{
+    Manager, PeerId,
+    behavior::{
+        InitiatorBehavior, InitiatorCommand, InitiatorEvent, PromotionBehavior, PromotionConfig,
+    },
+};
+use std::time::Duration;
+use tokio::select;
+use tracing::{Level, Span, instrument};
+
+pub type BlockBodyData = Vec<u8>;
+pub type RequestPort = gasket::messaging::InputPort<DownstreamRequest>;
+pub type SyncDataPort = gasket::messaging::OutputPort<ChainSyncEvent>;
+pub type BlockDataPort = gasket::messaging::OutputPort<BlockBodyData>;
+pub type NetworkMessage = pallas_network2::behavior::AnyMessage;
+pub type TcpInterface = pallas_network2::interface::TcpInterface<NetworkMessage>;
+pub type BlockRange = (Point, Point);
+
+const DEFAULT_HOUSEKEEPING_INTERVAL: Duration = Duration::from_secs(3);
+
+// preferring conversion via ad-hoc functions instead of implementing `From` to
+// avoid coupling other amaru crates with pallas-network2.
+
+fn amaru_to_pallas_point(point: &Point) -> pallas_network2::protocol::Point {
+    match point {
+        Point::Origin => pallas_network2::protocol::Point::Origin,
+        Point::Specific(slot, hash) => {
+            pallas_network2::protocol::Point::Specific(*slot, hash.clone())
+        }
+    }
+}
+
+fn pallas_to_amaru_point(point: &pallas_network2::protocol::Point) -> Point {
+    match point {
+        pallas_network2::protocol::Point::Origin => Point::Origin,
+        pallas_network2::protocol::Point::Specific(slot, hash) => {
+            Point::Specific(*slot, hash.clone())
+        }
+    }
+}
+
+fn pallas_to_amaru_peer(peer: &PeerId) -> amaru_kernel::peer::Peer {
+    amaru_kernel::peer::Peer::new(&peer.to_string())
+}
+
+#[derive(Default)]
+pub struct WorkUnit {
```

Comment:
Instead of stitching together `Option`s would it not make sense to have a composable enum, something like:

```
pub enum WorkUnit {
  Network(InitiatorCommand),
  Sync(ChainSyncEvent),
  Block(BlockBodyData),
  More(Box<WorkUnit>)
} 
```